### PR TITLE
First stab at nongenabi

### PR DESCRIPTION
**Pull request by wsmoses**
@vchuravy what are your thoughts on a design like this for supporting non-gen abi

### CODE REVIEW

File: src/compiler.jl [hunk 1]
```diff
@@ -6710,9 +6710,8 @@ end
 @inline remove_innerty(::Type{<:MixedDuplicated}) = MixedDuplicated
 @inline remove_innerty(::Type{<:BatchMixedDuplicated}) = MixedDuplicated
 
-@inline @generated function thunk(::Val{World}, ::Type{FA}, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{ModifiedBetween}, ::Val{ReturnPrimal}, ::Val{ShadowInit}, ::Type{ABI}) where {FA<:Annotation, A<:Annotation, TT, Mode, ModifiedBetween, width, ReturnPrimal, ShadowInit, World, ABI}   
+@inline function thunkbase(mi::Core.MethodInstance, ::Val{World}, ::Type{FA}, ::Type{A}, tt::Type{TT},::Val{Mode}, ::Val{width}, ::Val{ModifiedBetween}, ::Val{ReturnPrimal}, ::Val{ShadowInit}, ::Type{ABI}) where {FA<:Annotation, A<:Annotation, TT, Mode, ModifiedBetween, width, ReturnPrimal, ShadowInit, World, ABI}
```
Comment (vchuravy) [inline src/compiler.jl:6713]: It would be important that this function doesn't take world as an input.

Comment (wsmoses) [reply src/compiler.jl:6713]: Okay untested but done, regardless does this design make sense?

File: src/Enzyme.jl [hunk 2]
```diff
@@ -412,7 +414,9 @@ f(x) = x*x
     tt    = Tuple{map(T->eltype(Core.Typeof(T)), args)...}
     world = codegen_world_age(Core.Typeof(f.val), tt)
 
-    thunk = Enzyme.Compiler.thunk(Val(world), FA, RT, tt′, #=Mode=# Val(API.DEM_ForwardMode), Val(width),
+    opt_mi = RABI <: NonGenABI ? Compiler.fspec(eltype(FA), tt′) : Val(world)
```
Comment (vchuravy) [inline src/Enzyme.jl:417]: This observation of world is also an issue.

**Comment by vchuravy** [comment]
In order for code to relocatable you may not observe world as a compile time constant. You may look at the tls_world_age, but you may not cache that or put it into a Val.

File: src/Enzyme.jl (hunk 2 above)
Comment (wsmoses) [reply src/Enzyme.jl:417]: But it shouldn't occur for the nongen case, no ?

Or do I need to literally separate into two functionsfor the different ABI's?

Comment (vchuravy) [reply src/Enzyme.jl:417]: I would write it as:

```
    opt_mi = if RABI <: NonGenABI
        Compiler.fspec(eltype(FA), tt′)
    else
        Val(codegen_world_age(Core.Typeof(f.val), tt))
    end
```

To ensure that world is not accidentally captured. And then look at all uses of `codegen_world_age` and do the same.

**Pull request merged**